# 1. Знакомство с данными

## Импорт библиотек, модулей необходимых для решения задачи:

In [1]:
import wget 
import numpy as np # линейная алгебра
import pandas as pd # для анализа и предобработки данных
import matplotlib
import matplotlib.pyplot as plt # для визуализации
import seaborn as sns # для визуализации
import plotly.express as px # для визуализации
from sklearn import linear_model # линейные модели
from sklearn import metrics # метрики
import statistics
from collections import Counter
from functools import reduce
%matplotlib inline
plt.style.use('seaborn-v0_8')


from functools import reduce

## Импорт данных

In [2]:
# wget.download('https://github.com/rajathpi/assignment/raw/main/DAT465/nafld1.csv', 'D:\курс по data science\IDE\diplom_project\data_1')
# wget.download('https://github.com/rajathpi/assignment/raw/main/DAT465/nafld2.csv', 'D:\курс по data science\IDE\diplom_project\data_1')
# wget.download('https://github.com/rajathpi/assignment/raw/main/DAT465/nafld3.csv', 'D:\курс по data science\IDE\diplom_project\data_1')

## Загрузка данных и знакомство с ними

### Таблица № 1

In [3]:
nafld_1 = pd.read_csv('D:\курс по data science\IDE\diplom_project\data_1\\nafld1.csv')
display(nafld_1.head())
display(nafld_1.info()) # общая иинформация о датасете

Unnamed: 0  id  age  male  weight  height        bmi  case.id  futime  \
0        3631   1   57     0    60.0   163.0  22.690939  10630.0    6261   
1        8458   2   67     0    70.4   168.0  24.884028  14817.0     624   
2        6298   3   53     1   105.8   186.0  30.453537      3.0    1783   
3       15398   4   56     1   109.3   170.0  37.830100   6628.0    3143   
4       13261   5   68     1     NaN     NaN        NaN   1871.0    1836   

   status  
0       0  
1       0  
2       0  
3       0  
4       1

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 17549 entries, 0 to 17548
Data columns (total 10 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   Unnamed: 0  17549 non-null  int64  
 1   id          17549 non-null  int64  
 2   age         17549 non-null  int64  
 3   male        17549 non-null  int64  
 4   weight      12763 non-null  float64
 5   height      14381 non-null  float64
 6   bmi         12588 non-null  float64
 7   case.id     17518 non-null  float64
 8   futime      17549 non-null  int64  
 9   status      17549 non-null  int64  
dtypes: float64(4), int64(6)
memory usage: 1.3 MB


None

Таблица **'nafld_1'** содержит следущие признаки:
* **Unnamed: 0** - зашифрованный идентификтор пациента 
* **id** - идентификатор субъекта в исследовании;
* **age** - возраст при поступлении на обсследование;
* **male** - пол (0 = женский, 1 = мужской);
* **weight** - вес в кг;
* **height** - рост в см;
* **bmi** - индекс массы тела;
* **case.id** - идентификаатор сллучая НАЖБП, который соответствует этот объект;
* **futime** - прошедшее время до последнего наблюдения или смерти, в днях
* **status** - статус пациента (1 = мертвый, 0 = живой на момент последнего визита)

В данном датасете содержатся 10 столбцов, в которых данные в числовых форматах (int64 и float64), 17549 строк
В столбцах ['weight', 'height', 'bmi'] имеются пропуски

In [4]:
nafld_1[['age', 'weight', 'height', 'bmi', 'futime']].describe() # описательная статистика

age        weight        height           bmi        futime
count  17549.000000  12763.000000  14381.000000  12588.000000  17549.000000
mean      52.659639     86.353350    169.434949     30.073865   2410.600547
std       14.722515     22.239458     10.141421      7.085435   1573.017638
min       18.000000     33.400000    123.000000      9.207332      7.000000
25%       42.000000     70.000000    162.000000     25.136434   1132.000000
50%       53.000000     83.900000    169.000000     28.875884   2148.000000
75%       63.000000     99.200000    177.000000     33.710372   3353.000000
max       98.000000    181.700000    215.000000     84.395725   7268.000000

Исходя из описательной статитстики исходных данных видно следующее:
* средний возраст пациентов составляет 53 года, пациенты преимущественно взрослые, возраст колеблется от 18 до 98 лет 
* средний вес пациентов - 85 кг, диапозон показателей веса - 33-182 кг
* среднй рост пациентов - 169 см, рост самого низкого пациента составляет 123 см, самого высокого - 215 см
* Индекс массы тела (ИМТ) колеблется от 9,2 до 84,4 кг/м2, при интерпретации показателей ИМТ видим все возможные варианты от выраженного дефицита массы тела (min) до ожирения 3й степени, среднее значение ИМТ равен 30, что характерно для ожирения 1й степени
* среднее значение прошедшего времени до последнего наблюдения или смерти составляет 2410 дней или 6,6 лет, минимальное значение - 7 дней, макисмальное - 7268 дней или 20 лет

### Таблица № 2

In [5]:
nafld_2 = pd.read_csv('D:\курс по data science\IDE\diplom_project\data_1\\nafld2.csv')
display(nafld_2.head())
display(nafld_2.info()) # общая иинформация о датасете

Unnamed: 0  id  days  test  value
0      135077   1  -459   hdl   75.0
1      313143   1  -459  chol   75.0
2      135078   1   183   hdl   64.0
3      313144   1   183  chol   64.0
4      135079   1  2030   hdl   74.0

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 400123 entries, 0 to 400122
Data columns (total 5 columns):
 #   Column      Non-Null Count   Dtype  
---  ------      --------------   -----  
 0   Unnamed: 0  400123 non-null  int64  
 1   id          400123 non-null  int64  
 2   days        400123 non-null  int64  
 3   test        400123 non-null  object 
 4   value       400123 non-null  float64
dtypes: float64(1), int64(3), object(1)
memory usage: 15.3+ MB


None

Таблица **'nafld_2'** содержит следущие признаки:
* **Unnamed: 0** - 
* **id** - идентификатор субъекта;
* **days** - дни с даты ииндексирования;
* **test** - тип записанного значения;
* **value** - числовое значение

В данном датасете содержатся 5 столбцов, из которых 3 столбца в числовом формате 'int64', 1 столбец - в формате 'float64', и 1 столбец в формате 'object', 400123 строк
В столбцах не наблюдаются пропуски

In [6]:
display(nafld_2[['days', 'value']].describe())
display(nafld_2.describe(include=[object]))
display(nafld_2['test'].value_counts())

days          value
count  400123.000000  400123.000000
mean     -606.916616      63.768339
std      2415.953135      48.455567
min     -7651.000000       0.000000
25%     -2259.000000      40.000000
50%      -530.000000      51.000000
75%       987.000000      72.000000
max      7187.000000   13682.000000

test
count   400123
unique       6
top        hdl
freq    161259

hdl      161259
chol     161258
sbp       33430
dbp       33430
smoke      7008
fib4       3738
Name: test, dtype: int64

В столбце 'days' наблюдаются отрицательные значения,
описательная статистика столбца 'value' - пока что ничего не прояснила   

Анализ столбца 'test' выявил что имеются 6 уникальных значений, самое наиболее часто встречающеся - 'hdl'    

Могу предположить, что данные сокращения относятся к результатам лабораторных данных:
* ***hdl*** - значение холестерина липопротеинов высокой плотности в сыворотке крови;
* ***chol*** - значение общего холестерина;
* ***sbp*** - систолическое артиериальное давление;
* ***dbp*** - диастолическое артериальное давление;
* ***smoke*** - курит/не курит пациент;
* ***fib4*** - диагностический тест, предназначенный для определения наличия/отсутствия значимого фиброза печени у пациентов. 

### Таблица № 3

In [7]:
nafld_3 = pd.read_csv('D:\курс по data science\IDE\diplom_project\data_1\\nafld3.csv')
display(nafld_3.head())
display(nafld_3.info()) # общая иинформация о датасете

rownames  id  days         event
0         2   3     0         nafld
1         3   4 -1287           htn
2         4   4 -1287  dyslipidemia
3         5   4 -1226       ang/isc
4         7   4   -90       ang/isc

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 34340 entries, 0 to 34339
Data columns (total 4 columns):
 #   Column    Non-Null Count  Dtype 
---  ------    --------------  ----- 
 0   rownames  34340 non-null  int64 
 1   id        34340 non-null  int64 
 2   days      34340 non-null  int64 
 3   event     34340 non-null  object
dtypes: int64(3), object(1)
memory usage: 1.0+ MB


None

Таблица **'nafld_3'** содержит следущие признаки:
* **rownames** - 
* **id** - идентификатор субъекта;
* **days** - дни с даты ииндексирования;
* **event** - конечное событие которое произошло    


В данном датасете содержатся 4 столбца, которые все кроме одного столбца в числовом формате 'int64', 1 столбец - в формате 'object', 34340 строк

В столбцах не наблюдаются пропуски

In [8]:
display(nafld_3.describe(include=[object]))
display(nafld_3['event'].value_counts())

event
count          34340
unique            10
top     dyslipidemia
freq           10462

dyslipidemia      10462
htn                7097
nafld              3864
diabetes           3452
ang/isc            2235
stroke             2054
afib               1935
heart failure      1869
MI                 1199
cardiac arrest      173
Name: event, dtype: int64

В столбце 'event' таблицы 'nafld_3' выявлены 10 уникальных значений, наиболее часто встречающееся - 'dyslipidemia'

Можно предположить следующую расшифроку уникальныx значений столбца 'event':
* ***dyslipidemia*** - Дислипидемия (повышение уровней холестерина, триглицеридов (ТГ) или того и другого, либо низкий уровень холестерина липопротеинов высокой плотности (ХС-ЛПВП) в плазме, что способствует развитию атеросклероза)
* ***htn*** - Гипертония (устойчивое повышение систолического артериального давления в покое, диастолического артериального давления или того и другого)
* ***nafld*** - Неалкогольная жировая болезнь печени (избыточное накопление жира в гепатоцитах, клетках печени)
* ***diabetes*** - сахарный диабет (характеризуется нарушением секреции инсулина и той или иной степенью инсулинорезистентности, обусловливающими гипергликемию)
* ***ang/isc*** - угол наклона / периодическая самокатеризация (процесс регулярной катетеризации, которую пациент проводит самостоятельно для удаления мочи из мочевого пузыря)
* ***stroke*** - инсульт (острое нарушение кровообращения в головном мозге, сопровождающееся гибелью тканей и нарушением функций нервной системы)
* ***afib*** - Фибрилляция предсердий (аритмия (неритмичная работа сердца), то есть одна из форм сердечной патологии, при которой у человека обнаруживается нерегулярный пульс)
* ***heart failure*** - cердечная недостаточность (патологическое состояние, обусловленное несостоятельностью сердца как насоса, обеспечивающего адекватное кровообращение)
* ***MI*** - инфаркт миокарда (омертвление участка сердечной мышцы вследствие недостаточного ее кровоснабжения с развитием характерной клинической картины)
* ***cardiac arrest*** - Остановка сердца (происходит вследствие прекращения его механической активности, что приводит к отсутствию циркуляции кровотока. После остановки сердца прекращается кровоснабжение жизненно важных органов, что приводит к их кислородному голоданию.)

# 2. Подготовка данных



## Очистка данных
#### Дубликаты и пропуски таблицы №1

In [9]:
print(nafld_1.shape[0])
dupl_columns = list(nafld_1.columns)
mask = nafld_1.duplicated(subset=dupl_columns)
nafld_1_duplicates = nafld_1[mask]
print(f'Число найденных дубликатов: {nafld_1_duplicates.shape[0]}')
nafld_1_dedupped = nafld_1.drop_duplicates(subset=dupl_columns)
print(f'Результирующее число записей: {nafld_1_dedupped.shape[0]}')

17549
Число найденных дубликатов: 0
Результирующее число записей: 17549


In [10]:
null_nafld_1 = nafld_1_dedupped.isnull().sum()
display(null_nafld_1)

Unnamed: 0       0
id               0
age              0
male             0
weight        4786
height        3168
bmi           4961
case.id         31
futime           0
status           0
dtype: int64

Наблюдаем пропуски в столбцх ['weight','height', 'bmi', 'case.id'] \
Чтобы заполнить пропуски, необходимо выполнить следующее:
* Удалить строки в столбце ['case.id'] \
* Заполнить пропуски в столбцах ['weight', 'height'] медианным значением \
* В столбце ['bmi'] значением, рассчитанным по формуле для вычисления ИМТ

In [11]:
drop_nafld_1 = nafld_1_dedupped.copy()
drop_nafld_1 = drop_nafld_1.dropna(subset=['case.id'])
drop_nafld_1['weight'] = drop_nafld_1['weight'].fillna(drop_nafld_1['weight'].median())
drop_nafld_1['height'] = drop_nafld_1['height'].fillna(drop_nafld_1['height'].median())
drop_nafld_1['bmi'] = drop_nafld_1['bmi'].fillna(drop_nafld_1['weight']/((drop_nafld_1['height']/100)**2))
drop_nafld_1['bmi'] = drop_nafld_1['bmi'].round(1) # округлим значения ИМТ до 1 знака после запятой

null_1_nafld_1 = drop_nafld_1.isnull().sum()
display(null_1_nafld_1)

Unnamed: 0    0
id            0
age           0
male          0
weight        0
height        0
bmi           0
case.id       0
futime        0
status        0
dtype: int64

*Пропусков нет*

#### Дубликаты и пропуски таблицы №2

In [12]:
print(nafld_2.shape[0])
dupl_columns = list(nafld_2.columns)
mask = nafld_2.duplicated(subset=dupl_columns)
nafld_2_duplicates = nafld_2[mask]
print(f'Число найденных дубликатов: {nafld_2_duplicates.shape[0]}')
nafld_2_dedupped = nafld_2.drop_duplicates(subset=dupl_columns)
print(f'Результирующее число записей: {nafld_2_dedupped.shape[0]}')

400123
Число найденных дубликатов: 0
Результирующее число записей: 400123


In [13]:
null_nafld_2 = nafld_2_dedupped.isnull().sum()
display(null_nafld_2)

Unnamed: 0    0
id            0
days          0
test          0
value         0
dtype: int64

Таблица №2 чистая, нет пропусков, дубликатов

#### Дубликаты и пропуски таблицы №3

In [14]:
print(nafld_3.shape[0])
dupl_columns = list(nafld_3.columns)
mask = nafld_3.duplicated(subset=dupl_columns)
nafld_3_duplicates = nafld_3[mask]
print(f'Число найденных дубликатов: {nafld_3_duplicates.shape[0]}')
nafld_3_dedupped = nafld_3.drop_duplicates(subset=dupl_columns)
print(f'Результирующее число записей: {nafld_3_dedupped.shape[0]}')

34340
Число найденных дубликатов: 0
Результирующее число записей: 34340


In [15]:
null_nafld_3 = nafld_3_dedupped.isnull().sum()
display(null_nafld_3)

rownames    0
id          0
days        0
event       0
dtype: int64

Таблица №3 чистая, нет пропусков, дубликатов

## Получение новых признаков
#### Таблица №1

Можно удалить столбцы 'Unnamed: 0','case.id','futime' так как не несет смысловой нагрузки

In [16]:
drop_nafld_1 = drop_nafld_1.drop(['Unnamed: 0','case.id','futime'], axis=1)
drop_nafld_1.head(2)

id  age  male  weight  height   bmi  status
0   1   57     0    60.0   163.0  22.7       0
1   2   67     0    70.4   168.0  24.9       0

По результатам ИМТ можно сделать следующие категории:

In [17]:
drop_nafld_1.loc[drop_nafld_1['bmi'].between(0, 16, inclusive='both'),'bmi_category'] = 'severe underweight'
drop_nafld_1.loc[drop_nafld_1['bmi'].between(16.1, 18.4, inclusive='both'),'bmi_category'] = 'underweight'
drop_nafld_1.loc[drop_nafld_1['bmi'].between(18.5, 24.9, inclusive='both'),'bmi_category'] = 'healthy'
drop_nafld_1.loc[drop_nafld_1['bmi'].between(25.0, 29.9, inclusive='both'),'bmi_category'] = 'pre-obesity'
drop_nafld_1.loc[drop_nafld_1['bmi'].between(30.0, 34.9, inclusive='both'),'bmi_category'] = 'obese (class I)'
drop_nafld_1.loc[drop_nafld_1['bmi'].between(35.0, 39.9, inclusive='both'),'bmi_category'] = 'obese (class II)'
drop_nafld_1.loc[drop_nafld_1['bmi'].between(40.0, 100, inclusive='both'),'bmi_category'] = 'obese (class III)'

drop_nafld_1['bmi_category'].value_counts()

pre-obesity           8029
obese (class I)       3540
healthy               3127
obese (class II)      1556
obese (class III)     1155
underweight             89
severe underweight      22
Name: bmi_category, dtype: int64

#### Таблица №2

In [18]:
display(nafld_2.head(3))

Unnamed: 0  id  days  test  value
0      135077   1  -459   hdl   75.0
1      313143   1  -459  chol   75.0
2      135078   1   183   hdl   64.0

Также и в таблице №1 можно удалить столбец 'Unnamed: 0' и 'days', так как не несет смысловой нагрузки

In [19]:
drop_nafld_2 = nafld_2.copy()
drop_nafld_2 = drop_nafld_2.drop(['Unnamed: 0', 'days'], axis=1)
drop_nafld_2.head(3)

id  test  value
0   1   hdl   75.0
1   1  chol   75.0
2   1   hdl   64.0

Для каждого теста имеются референсные значения диапазона норм показателей по каждому тесту

In [20]:
drop_nafld_2['test'].value_counts()

hdl      161259
chol     161258
sbp       33430
dbp       33430
smoke      7008
fib4       3738
Name: test, dtype: int64

Референсные значения диапазона норм показателей по тесту **hdl**, демонстрирующий количество холестерина липопротеинов высокой плотности в сыворотке крови

In [26]:
hdl = drop_nafld_2.loc[drop_nafld_2['test'] == 'hdl']
hdl_group = hdl.groupby('id')['value'].apply(lambda x: ', '.join(x.astype(str))).reset_index()
hdl_group['value'] = hdl_group['value'].str.split(', ').apply(lambda x: [float(i) for i in x])

hdl_group['MedianaHdl'] = hdl_group['value'].apply(lambda x: statistics.median(x) if x else 0)

   
def get_hdl(value):
    if value < 40.0:
        return 'low'
    elif 40.0 <= value <= 59.0:
        return 'normal'
    else:
        return 'ideal'
    

hdl_group.loc[:, 'result_hdl_mediana'] = hdl_group['MedianaHdl'].apply(get_hdl)
hdl_group.head(3)


id                                 value  MedianaHdl result_hdl_mediana
0   1        [75.0, 64.0, 74.0, 50.0, 79.0]        74.0              ideal
1   2  [61.0, 56.0, 87.0, 69.0, 86.0, 86.0]        77.5              ideal
2   3                    [52.0, 63.0, 52.0]        52.0             normal

Референсные значения диапазона норм показателей по тесту **chol**, демонстрирующий количество общего холестерина в сыворотке крови

In [27]:
chol = drop_nafld_2.loc[drop_nafld_2['test'] == 'chol']
chol_group = chol.groupby('id')['value'].apply(lambda x: ', '.join(x.astype(str))).reset_index()
chol_group['value'] = chol_group['value'].str.split(', ').apply(lambda x: [float(i) for i in x])

chol_group['MedianaChol'] = chol_group['value'].apply(lambda x: statistics.median(x) if x else 0)

def get_chol(value):
    if value < 40.0:
        return 'low'
    elif 40.0 <= value <= 59.0:
        return 'normal'
    else:
        return 'ideal'

chol_group.loc[:, 'result_chol_mediana'] = chol_group['MedianaChol'].apply(get_hdl)

chol_group.head(3)

id                                 value  MedianaChol result_chol_mediana
0   1        [75.0, 64.0, 74.0, 50.0, 79.0]         74.0               ideal
1   2  [61.0, 56.0, 87.0, 69.0, 86.0, 86.0]         77.5               ideal
2   3                    [52.0, 63.0, 52.0]         52.0              normal

Референсные значения диапазона норм показателей по тесту **sbp**, демонстрирующий показатель систолического артиериального давления

In [28]:
sbp = drop_nafld_2.loc[drop_nafld_2['test'] == 'sbp']
sbp_group = sbp.groupby('id')['value'].apply(lambda x: ', '.join(x.astype(str))).reset_index()
sbp_group['value'] = sbp_group['value'].str.split(', ').apply(lambda x: [float(i) for i in x])

def get_sbp(value):
    if 40 < value < 105:
        return 'low'
    elif 105 <= value < 120:
        return 'ideal'
    elif 120 <= value < 130:
        return 'normal'
    elif 130 <= value < 140:
        return 'high normal'
    else:
        return 'hypertension'

# sbp_group['MeanSbp'] = sbp_group['value'].apply(lambda x: sum(x) / len(x) if x else 0)
sbp_group['MedianaSbp'] = sbp_group['value'].apply(lambda x: statistics.median(x) if x else 0)

# sbp_group.loc[:, 'result_sbp_mean'] = sbp_group['MeanSbp'].apply(get_sbp)
sbp_group.loc[:, 'result_sbp_mediana'] = sbp_group['MedianaSbp'].apply(get_sbp)

sbp_group.head(3)

id                                       value  MedianaSbp  \
0   3                                     [135.0]       135.0   
1   4                       [142.0, 148.0, 109.0]       142.0   
2   9  [157.0, 132.0, 110.0, 118.0, 122.0, 136.0]       127.0   

  result_sbp_mediana  
0        high normal  
1       hypertension  
2             normal

Референсные значения диапазона норм показателей по тесту **dbp**, демонстрирующий показатель диастолического артиериального давления

In [29]:
dbp = drop_nafld_2.loc[drop_nafld_2['test'] == 'dbp']
dbp_group = dbp.groupby('id')['value'].apply(lambda x: ', '.join(x.astype(str))).reset_index()
dbp_group['value'] = dbp_group['value'].str.split(', ').apply(lambda x: [float(i) for i in x])

def get_dbp(value):
    if 40 < value < 60:
        return 'low'
    elif 60 <= value < 80:
        return 'ideal'
    elif 80 <= value < 85:
        return 'normal'
    elif 85 <= value < 90:
        return 'high normal'
    else:
        return 'hypertension'

# dbp_group['MeanDbp'] = dbp_group['value'].apply(lambda x: sum(x) / len(x) if x else 0)
dbp_group['MedianaDbp'] = dbp_group['value'].apply(lambda x: statistics.median(x) if x else 0)

# dbp_group.loc[:, 'result_dbp_mean'] = dbp_group['MeanDbp'].apply(get_dbp)
dbp_group.loc[:, 'result_dbp_mediana'] = dbp_group['MedianaDbp'].apply(get_dbp)
dbp_group.head(3)

id                                       value  MedianaDbp  \
0   3                                     [135.0]       135.0   
1   4                       [142.0, 148.0, 109.0]       142.0   
2   9  [157.0, 132.0, 110.0, 118.0, 122.0, 136.0]       127.0   

  result_dbp_mediana  
0       hypertension  
1       hypertension  
2       hypertension

* ***hdl*** - значение холестерина липопротеинов высокой плотности в сыворотке крови;
* ***chol*** - значение общего холестерина;
* ***sbp*** - систолическое артиериальное давление;
* ***dbp*** - диастолическое артериальное давление;
* ***smoke*** - курит/не курит пациент;
* ***fib4*** - диагностический тест, предназначенный для определения наличия/отсутствия значимого фиброза печени у пациентов. 

Референсные значения диапазона норм показателей по диагностическому тесту **fib4**, предназначенный для определения наличия/отсутствия значимого фиброза печени у пациентов

In [30]:
fib4 = drop_nafld_2.loc[drop_nafld_2['test'] == 'fib4']

def get_fib4(value):
    if value < 1.3:
        return 'low'
    elif 1.3 <= value <= 2.67:
        return 'risk'
    else:
        return 'cirrosis'

fib4.loc[:, 'result_fib4'] = fib4['value'].apply(get_fib4)

fib4.head(3)

C:\Users\Елена\AppData\Local\Temp\ipykernel_24512\2337041443.py:11: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  fib4.loc[:, 'result_fib4'] = fib4['value'].apply(get_fib4)


id  test     value result_fib4
48    4  fib4  4.554200    cirrosis
213  10  fib4  4.719419    cirrosis
258  12  fib4  3.175664    cirrosis

Показатель "smoke" - курит/не курит пациент

In [31]:
smoke = drop_nafld_2.loc[drop_nafld_2['test'] == 'smoke']

def get_smoke(value):
    if value == 1:
        return 'smoke'
    else:
        return 'no smoke'

smoke['result_smoke'] = smoke['value'].apply(get_smoke)

smoke.head(3)

C:\Users\Елена\AppData\Local\Temp\ipykernel_24512\675107422.py:9: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  smoke['result_smoke'] = smoke['value'].apply(get_smoke)


id   test  value result_smoke
400  19  smoke    1.0        smoke
403  19  smoke    0.0     no smoke
488  25  smoke    0.0     no smoke

Теперь объединим получившиеся датафреймы с референсными значениями в один датафрейм

In [33]:
nafld_2_groupped = drop_nafld_2.groupby('id')['test'].apply(lambda x: ', '.join(x)).reset_index()

def remove_duplicates(text):
    text = text.split(', ') # split text into words
    text = Counter(text) # count the occurrence of each word
    text = ", ".join(text.keys()) # join unique words back into a string
    return text

nafld_2_groupped['test'] = nafld_2_groupped['test'].apply(remove_duplicates)
nafld_2_groupped.head(3)

id                 test
0   1            hdl, chol
1   2            hdl, chol
2   3  hdl, chol, sbp, dbp

In [34]:
dataframes = [nafld_2_groupped, hdl_group, chol_group, sbp_group, dbp_group, fib4, smoke]

nafld_2_join = reduce(lambda left,right: pd.merge(left,right,on='id', how='outer'), dataframes)
nafld_2_join.head(3)

C:\Users\Елена\AppData\Local\Temp\ipykernel_24512\3002727236.py:3: FutureWarning: Passing 'suffixes' which cause duplicate columns {'value_x'} in the result is deprecated and will raise a MergeError in a future version.
  nafld_2_join = reduce(lambda left,right: pd.merge(left,right,on='id', how='outer'), dataframes)
C:\Users\Елена\AppData\Local\Temp\ipykernel_24512\3002727236.py:3: FutureWarning: Passing 'suffixes' which cause duplicate columns {'value_x'} in the result is deprecated and will raise a MergeError in a future version.
  nafld_2_join = reduce(lambda left,right: pd.merge(left,right,on='id', how='outer'), dataframes)


id               test_x                               value_x  MedianaHdl  \
0   1            hdl, chol        [75.0, 64.0, 74.0, 50.0, 79.0]        74.0   
1   2            hdl, chol  [61.0, 56.0, 87.0, 69.0, 86.0, 86.0]        77.5   
2   3  hdl, chol, sbp, dbp                    [52.0, 63.0, 52.0]        52.0   

  result_hdl_mediana                               value_y  MedianaChol  \
0              ideal        [75.0, 64.0, 74.0, 50.0, 79.0]         74.0   
1              ideal  [61.0, 56.0, 87.0, 69.0, 86.0, 86.0]         77.5   
2             normal                    [52.0, 63.0, 52.0]         52.0   

  result_chol_mediana  value_x  MedianaSbp result_sbp_mediana  value_y  \
0               ideal      NaN         NaN                NaN      NaN   
1               ideal      NaN         NaN                NaN      NaN   
2              normal  [135.0]       135.0        high normal  [135.0]   

   MedianaDbp result_dbp_mediana test_y  value_x result_fib4 test  value_y  \
0         NaN                NaN    NaN      NaN         NaN  NaN      NaN   
1         NaN                NaN    NaN      NaN         NaN  NaN      NaN   
2       135.0       hypertension    NaN      NaN         NaN  NaN      NaN   

  result_smoke  
0          NaN  
1          NaN  
2          NaN

In [35]:
nafld_2_join.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 19457 entries, 0 to 19456
Data columns (total 20 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   id                   19457 non-null  int64  
 1   test_x               19457 non-null  object 
 2   value_x              18828 non-null  object 
 3   MedianaHdl           18828 non-null  float64
 4   result_hdl_mediana   18828 non-null  object 
 5   value_y              18827 non-null  object 
 6   MedianaChol          18827 non-null  float64
 7   result_chol_mediana  18827 non-null  object 
 8   value_x              9955 non-null   object 
 9   MedianaSbp           9955 non-null   float64
 10  result_sbp_mediana   9955 non-null   object 
 11  value_y              9955 non-null   object 
 12  MedianaDbp           9955 non-null   float64
 13  result_dbp_mediana   9955 non-null   object 
 14  test_y               5180 non-null   object 
 15  value_x              5180 non-null  

Удалим лишние столбцы

In [36]:
nafld_2_join = nafld_2_join.drop(['value_y', 'value_x', 'test_y', 'test_x', 'test'], axis=1)
nafld_2_join.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 19457 entries, 0 to 19456
Data columns (total 11 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   id                   19457 non-null  int64  
 1   MedianaHdl           18828 non-null  float64
 2   result_hdl_mediana   18828 non-null  object 
 3   MedianaChol          18827 non-null  float64
 4   result_chol_mediana  18827 non-null  object 
 5   MedianaSbp           9955 non-null   float64
 6   result_sbp_mediana   9955 non-null   object 
 7   MedianaDbp           9955 non-null   float64
 8   result_dbp_mediana   9955 non-null   object 
 9   result_fib4          5180 non-null   object 
 10  result_smoke         7058 non-null   object 
dtypes: float64(4), int64(1), object(6)
memory usage: 1.8+ MB


In [37]:
nafld_2_join.head(3)

id  MedianaHdl result_hdl_mediana  MedianaChol result_chol_mediana  \
0   1        74.0              ideal         74.0               ideal   
1   2        77.5              ideal         77.5               ideal   
2   3        52.0             normal         52.0              normal   

   MedianaSbp result_sbp_mediana  MedianaDbp result_dbp_mediana result_fib4  \
0         NaN                NaN         NaN                NaN         NaN   
1         NaN                NaN         NaN                NaN         NaN   
2       135.0        high normal       135.0       hypertension         NaN   

  result_smoke  
0          NaN  
1          NaN  
2          NaN

#### Таблица №3


In [38]:
nafld_3.head(2)

rownames  id  days  event
0         2   3     0  nafld
1         3   4 -1287    htn

Удалить столбец 'rownames'/
Дни преобразовть в года

In [39]:
drop_nafld_3 = nafld_3.copy()
drop_nafld_3 = drop_nafld_3.drop('rownames', axis=1)
drop_nafld_3['years'] = round(drop_nafld_3['days']/365, 2)
drop_nafld_3.head(3)

id  days         event  years
0   3     0         nafld   0.00
1   4 -1287           htn  -3.53
2   4 -1287  dyslipidemia  -3.53

In [40]:
drop_nafld_3['id'].value_counts()

17430    14
1682     13
3238     13
3842     13
3784     12
         ..
12730     1
5952      1
12728     1
5953      1
3         1
Name: id, Length: 12454, dtype: int64

In [77]:
# drop_nafld_3.loc[drop_nafld_3['id'] == 3238]

In [78]:
# display(drop_nafld_1.loc[drop_nafld_1['id'] == 15446])
# display(drop_nafld_2.loc[drop_nafld_2['id'] == 15446])
# display(drop_nafld_3.loc[drop_nafld_3['id'] == 15446])

Выполним кодировку категориального признака "event" - диагнозов, имеющихся у пациента

In [42]:
nafld_3_grouped = drop_nafld_3.groupby('id')['event'].apply(lambda x: ', '.join(x)).reset_index()
nafld_3_grouped.head(3)

id                                event
0   3                                nafld
1   4  htn, dyslipidemia, ang/isc, ang/isc
2   5                                  htn

In [44]:
nafld_3_grouped['event_list'] = nafld_3_grouped['event'].apply(lambda x: x.split(', '))
nafld_3_grouped['nafld'] = nafld_3_grouped['event_list'].apply(lambda x: 1 if 'nafld' in x else 0)
nafld_3_grouped['dyslipidemia'] = nafld_3_grouped['event_list'].apply(lambda x: 1 if 'dyslipidemia' in x else 0)
nafld_3_grouped['ang/isc'] = nafld_3_grouped['event_list'].apply(lambda x: 1 if 'ang/isc' in x else 0)
nafld_3_grouped['diabetes'] = nafld_3_grouped['event_list'].apply(lambda x: 1 if 'diabetes' in x else 0)
nafld_3_grouped['stroke'] = nafld_3_grouped['event_list'].apply(lambda x: 1 if 'stroke' in x else 0)
nafld_3_grouped['afib'] = nafld_3_grouped['event_list'].apply(lambda x: 1 if 'afib' in x else 0)
nafld_3_grouped['heart failure'] = nafld_3_grouped['event_list'].apply(lambda x: 1 if 'heart failure' in x else 0)
nafld_3_grouped['MI'] = nafld_3_grouped['event_list'].apply(lambda x: 1 if 'MI' in x else 0)
nafld_3_grouped['cardiac arrest'] = nafld_3_grouped['event_list'].apply(lambda x: 1 if 'cardiac arrest' in x else 0)
nafld_3_grouped['htn'] = nafld_3_grouped['event_list'].apply(lambda x: 1 if 'htn' in x else 0)

nafld_3_grouped = nafld_3_grouped.drop('event_list', axis=1)
nafld_3_grouped.head(3)

id                                event  nafld  dyslipidemia  ang/isc  \
0   3                                nafld      1             0        0   
1   4  htn, dyslipidemia, ang/isc, ang/isc      0             1        1   
2   5                                  htn      0             0        0   

   diabetes  stroke  afib  heart failure  MI  cardiac arrest  htn  
0         0       0     0              0   0               0    0  
1         0       0     0              0   0               0    1  
2         0       0     0              0   0               0    1

Сохранить преобразованные датафреймы

In [45]:
drop_nafld_1.to_csv('D:\курс по data science\IDE\diplom_project\data_2\processed_nafld_1.csv', index=False)
nafld_2_join.to_csv('D:\курс по data science\IDE\diplom_project\data_2\processed_nafld_2.csv', index=False)
nafld_3_grouped.to_csv('D:\курс по data science\IDE\diplom_project\data_2\processed_nafld_3.csv', index=False)